File: output/cpp/tests/03_transition.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>
#include <functional>


#include <iostream>
#include <string>
#include <cassert>

class WithTransitionFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    WithTransitionFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class WithTransitionFrameContext {
public:
    WithTransitionFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    WithTransitionFrameContext(WithTransitionFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class WithTransitionCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<WithTransitionFrameEvent> forward_event;
    std::unique_ptr<WithTransitionCompartment> parent_compartment;

    explicit WithTransitionCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<WithTransitionCompartment> clone() const {
        auto c = std::make_unique<WithTransitionCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class WithTransition {
private:
    std::vector<std::unique_ptr<WithTransitionCompartment>> _state_stack;
    std::unique_ptr<WithTransitionCompartment> __compartment;
    std::unique_ptr<WithTransitionCompartment> __next_compartment;
    std::vector<WithTransitionFrameContext> _context_stack;

    void __kernel(WithTransitionFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            WithTransitionFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                WithTransitionFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    WithTransitionFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(WithTransitionFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "First") {
            _state_First(__e);
        } else if (state_name == "Second") {
            _state_Second(__e);
        }
    }

    void __transition(std::unique_ptr<WithTransitionCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Second(WithTransitionFrameEvent& __e) {
        if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("Second"));
            return;;
        } else if (__e._message == "next") {
            printf("Transitioning: Second -> First\n");
            auto __new_compartment = std::make_unique<WithTransitionCompartment>("First");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        }
    }

    void _state_First(WithTransitionFrameEvent& __e) {
        if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("First"));
            return;;
        } else if (__e._message == "next") {
            printf("Transitioning: First -> Second\n");
            auto __new_compartment = std::make_unique<WithTransitionCompartment>("Second");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        }
    }

public:
    WithTransition() {
        __compartment = std::make_unique<WithTransitionCompartment>("First");
        WithTransitionFrameEvent __frame_event("$>");
        WithTransitionFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void next() {
        WithTransitionFrameEvent __e("next");
        WithTransitionFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    std::string get_state() {
        WithTransitionFrameEvent __e("get_state");
        WithTransitionFrameContext __ctx(std::move(__e), std::any(std::string()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::string>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }
};

int main() {
    printf("=== Test 03: State Transitions ===\n");
    WithTransition s;

    std::string state = s.get_state();
    if (state != "First") {
        printf("FAIL: Expected 'First', got '%s'\n", state.c_str());
        assert(false);
    }
    printf("Initial state: %s\n", state.c_str());

    s.next();
    state = s.get_state();
    if (state != "Second") {
        printf("FAIL: Expected 'Second', got '%s'\n", state.c_str());
        assert(false);
    }
    printf("After first next(): %s\n", state.c_str());

    s.next();
    state = s.get_state();
    if (state != "First") {
        printf("FAIL: Expected 'First', got '%s'\n", state.c_str());
        assert(false);
    }
    printf("After second next(): %s\n", state.c_str());

    printf("PASS: State transitions work correctly\n");
    return 0;
}

File: output/go/tests/03_transition.go

package main

import "fmt"
import "os"

type WithTransitionFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type WithTransitionFrameContext struct {
    _event  WithTransitionFrameEvent
    _return any
    _data   map[string]any
}

type WithTransitionCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *WithTransitionFrameEvent
    parentCompartment *WithTransitionCompartment
}

func newWithTransitionCompartment(state string) *WithTransitionCompartment {
    return &WithTransitionCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *WithTransitionCompartment) copy() *WithTransitionCompartment {
    nc := &WithTransitionCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type WithTransition struct {
    _state_stack []*WithTransitionCompartment
    __compartment *WithTransitionCompartment
    __next_compartment *WithTransitionCompartment
    _context_stack []WithTransitionFrameContext
}

func NewWithTransition() *WithTransition {
    s := &WithTransition{}
    s._state_stack = make([]*WithTransitionCompartment, 0)
    s._context_stack = make([]WithTransitionFrameContext, 0)
    s.__compartment = newWithTransitionCompartment("First")
    s.__next_compartment = nil
    __frame_event := WithTransitionFrameEvent{_message: "$>", _parameters: nil}
    __ctx := WithTransitionFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *WithTransition) __kernel(__e *WithTransitionFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &WithTransitionFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &WithTransitionFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &WithTransitionFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *WithTransition) __router(__e *WithTransitionFrameEvent) {
    switch s.__compartment.state {
    case "First":
        s._state_First(__e)
    case "Second":
        s._state_Second(__e)
    }
}

func (s *WithTransition) __transition(next *WithTransitionCompartment) {
    s.__next_compartment = next
}

func (s *WithTransition) Next() {
    __e := WithTransitionFrameEvent{_message: "Next"}
    __ctx := WithTransitionFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *WithTransition) GetState() string {
    __e := WithTransitionFrameEvent{_message: "GetState"}
    __ctx := WithTransitionFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *WithTransition) _state_Second(__e *WithTransitionFrameEvent) {
    if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "Second"
        return
    } else if __e._message == "Next" {
        __compartment := newWithTransitionCompartment("First")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func (s *WithTransition) _state_First(__e *WithTransitionFrameEvent) {
    if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "First"
        return
    } else if __e._message == "Next" {
        __compartment := newWithTransitionCompartment("Second")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func main() {
	fmt.Println("=== Test 03: State Transitions ===")
	s := NewWithTransition()

	state := s.GetState()
	if state != "First" {
		fmt.Printf("FAIL: Expected 'First', got '%s'\n", state)
		os.Exit(1)
	}
	fmt.Printf("Initial state: %s\n", state)

	s.Next()
	state = s.GetState()
	if state != "Second" {
		fmt.Printf("FAIL: Expected 'Second', got '%s'\n", state)
		os.Exit(1)
	}
	fmt.Printf("After first Next(): %s\n", state)

	s.Next()
	state = s.GetState()
	if state != "First" {
		fmt.Printf("FAIL: Expected 'First', got '%s'\n", state)
		os.Exit(1)
	}
	fmt.Printf("After second Next(): %s\n", state)

	fmt.Println("PASS: State transitions work correctly")
}
